### PR TITLE
Add NodeWalker for walking nodes without storing them.

### PR DESCRIPTION
**Pull request by jspam**
This avoid allocating and discarding an ArrayBuffer of nodes every time.

### CODE REVIEW

File: src/main/scala/mjis/opt/FirmExtensions.scala [hunk 1]
```diff
@@ -33,11 +33,11 @@ object FirmExtensions {
     }
 
     def getInductionVariables: Seq[InductionVariable] = {
-      NodeCollector.fromWalk(g.walk).map(node => node match {
+      NodeCollector.fromWalk(g.walk).flatMap(node => node match {
```
Comment (TehMillhouse) [inline src/main/scala/mjis/opt/FirmExtensions.scala]: Just curious... what's the reason we can't use the `NodeWalker` here?

**Comment by TehMillhouse** [comment]
A suggestion: Maybe put `NodeCounter` and `NodeWalker` in `GraphExt`? In general: LGTM

**Comment by jspam** [comment]
Like this?

**Comment by TehMillhouse** [comment]
Exactly like this :)
Feel free to merge if you're happy with the state.

**Pull request merged**